### PR TITLE
fat_fs: Add support for variable sector size

### PR DESCRIPTION
**Pull request by mrfuchs**
Add support for variable sector sizes by exposing _MAX_SS as
CONFIG_FS_FATFS_MAX_SS to Kconfig and adding the GET_SECTOR_SIZE IOCTL.

Signed-off-by: Markus Fuchs <markus.fuchs@de.sauter-bc.com>

### CODE REVIEW

File: include/ffconf.h [hunk 1]
```diff
@@ -189,7 +189,11 @@
 
 
 #define	_MIN_SS		512
+#if defined(CONFIG_FS_FATFS_MAX_SS)
+#define	_MAX_SS		CONFIG_FS_FATFS_MAX_SS
```
Comment (de-nordic) [inline include/ffconf.h:193]: Update the README list of exported config parameters, similar to the PR; https://github.com/zephyrproject-rtos/fatfs/pull/4

File: zfs_diskio.c [hunk 2]
```diff
@@ -99,6 +99,8 @@ DRESULT disk_ioctl(BYTE pdrv, BYTE cmd, void *buff)
 {
 	int ret =  RES_OK;
 
+	uint32_t sector_size;
```
[the remark this answers is not in the document]
Comment (mrfuchs) [reply zfs_diskio.c]: Initialized `sector_size` to 0.

File: include/ffconf.h (hunk 1 above)
Comment (mrfuchs) [reply include/ffconf.h:193]: Updated README.

**Comment by carlescufi** [comment]
@de-nordic please revisit
@mrfuchs please rebase

File: zfs_diskio.c [hunk 3]
```diff
@@ -113,7 +115,20 @@ DRESULT disk_ioctl(BYTE pdrv, BYTE cmd, void *buff)
...
+		break;
+
+	case GET_SECTOR_SIZE:
+		/* Zephyr's DISK_IOCTL_GET_SECTOR_SIZE returns sector size as a
+		 * 32-bit number while FatFS's GET_SECTOR_SIZE is supposed to
+		 * return a 16-bit number.
+		 */
+		if (disk_access_ioctl(pdrv_str[pdrv],
+				DISK_IOCTL_GET_SECTOR_SIZE, &sector_size) == 0) {
+			*(uint16_t *)buff = (uint16_t)sector_size;
```
Comment (de-nordic) [inline zfs_diskio.c:128]: I think this requires some assertion or check of higher 16bits against 0.

Comment (pabigot) [reply zfs_diskio.c:128]: Agreed; add `(sector_size == (uint16_t)sector_size)` as a secondary condition.

Comment (mrfuchs) [reply zfs_diskio.c:129]: Added check. Thanks!

File: zfs_diskio.c [hunk 4]
```diff
@@ -97,7 +97,9 @@ DRESULT disk_write(BYTE pdrv, const BYTE *buff, DWORD sector, UINT count)
 
 DRESULT disk_ioctl(BYTE pdrv, BYTE cmd, void *buff)
 {
-	int ret =  RES_OK;
+	int ret = RES_OK;
+
```
Comment (pabigot) [inline zfs_diskio.c:101]: remove empty line between declarations

**Review by pabigot — CHANGES_REQUESTED** [review]
One style nitpick, then looks good.

File: zfs_diskio.c (hunk 4 above)
Comment (mrfuchs) [reply zfs_diskio.c]: Removed empty line.

**Review by pabigot — APPROVED** [review]

**Review by de-nordic — APPROVED** [review]

**Pull request merged**